src/server/hset_family.cc

```diff
@@ -1020,24 +1050,90 @@ void HSetFamily::HRandField(CmdArgList args, ConnectionContext* cntx) {
     StringVec str_vec;
 
     if (pv.Encoding() == kEncodingStrMap2) {
-      // TODO: to create real random logic.
       StringMap* string_map = (StringMap*)pv.RObjPtr();
+      if (args.size() == 1) {
+        sds key;
+        string_map->RandomPair(&key, nullptr);
+        // sds key = string_map->begin()->first;
+        str_vec.emplace_back(key, sdslen(key));
+      } else {
+        size_t actual_count =
+            (count >= 0) ? std::min(size_t(count), string_map->Size()) : abs(count);
+        sds *keys = nullptr, *vals = nullptr;
+        keys = new sds[actual_count];
+        if (!with_values.empty())
+          vals = new sds[actual_count];
+        if (count >= 0) {
+          if (!with_values.empty()) {
+            string_map->RandomPairsUnique(actual_count, keys, vals);
+          } else {
+            string_map->RandomPairsUnique(actual_count, keys, nullptr);
+          }
+        } else {
+          if (!with_values.empty()) {
+            string_map->RandomPairs(actual_count, keys, vals);
+          } else {
+            string_map->RandomPairs(actual_count, keys, nullptr);
+          }
+        }
+        for (size_t i = 0; i < actual_count; ++i) {
+          str_vec.emplace_back(keys[i], sdslen(keys[i]));
+          if (!with_values.empty()) {
+            str_vec.emplace_back(vals[i], sdslen(vals[i]));
+          }
+        }
+        delete[] keys;
+        if (!with_values.empty())
+          delete[] vals;
+      }
 
-      sds key = string_map->begin()->first;
-      str_vec.emplace_back(key, sdslen(key));
     } else if (pv.Encoding() == kEncodingListPack) {
       uint8_t* lp = (uint8_t*)pv.RObjPtr();
       size_t lplen = lpLength(lp);
       CHECK(lplen > 0 && lplen % 2 == 0);
-
       size_t hlen = lplen / 2;
-      listpackEntry key;
-
-      lpRandomPair(lp, hlen, &key, NULL);
-      if (key.sval) {
-        str_vec.emplace_back(reinterpret_cast<char*>(key.sval), key.slen);
+      if (args.size() == 1) {
+        listpackEntry key;
+        lpRandomPair(lp, hlen, &key, NULL);
+        StrVecEmplaceBack(str_vec, key);
       } else {
-        str_vec.emplace_back(absl::StrCat(key.lval));
+        size_t actual_count = (count >= 0) ? std::min(size_t(count), hlen) : abs(count);
+        listpackEntry *keys = nullptr, *vals = nullptr;
+        keys = new listpackEntry[actual_count];
+        if (!with_values.empty())
+          vals = new listpackEntry[actual_count];
+
+        // count has been specified.
+        if (count >= 0) {
+          // always returns unique entries.
+          if (!with_values.empty()) {
+            // WITHVALUES is specified.
+            // randomly returns actual_count unique keys and values.
+            lpRandomPairsUnique(lp, actual_count, keys, vals);
+          } else {
+            lpRandomPairsUnique(lp, actual_count, keys, NULL);
```

Comment:
same here, its already null, so no if is needed

Reply:
Nice catch! This makes my suggestion with the `body` lambda even more plausible/simple. +1

Reply:
@dranikpg  see my comments/questions above.